### ISSUE TITLE
adding custom currency

### ISSUE THREAD

**Issue by farshadff** [bagisto/bagisto#1814]
i think not all currencies have 2 zeros like my own and i want to remove them . what if i want to add the utf8 code which is not possible now and when i change it manually from database the price wont show up any more
thanks

**Comment by farshadff** [comment]
any idea how to achieve this ?

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

We have used NumberFormatter to format currency for which we have created a facades called core.
First you need to create a package where you need to create a facades , now your this facades will override our core facades. like - 

`
        $loader = AliasLoader::getInstance();

        $loader->alias('core', Core::class);

        $this->app->singleton('core', function () {
            return new core();
        });

        $this->app->bind('core', 'Your\Facdes\path');
` 
IF you see our facades, there you can find  **formatPrice** function , you need to override this function.

`
 public function formatPrice($price, $currencyCode)
    {
        if (is_null($price))
            $price = 0;

        $formater = new \NumberFormatter( app()->getLocale(), \NumberFormatter::CURRENCY );

        $formater->setAttribute( $formater::FRACTION_DIGITS, 0 );

        if ($symbol = $this->getCurrentCurrency()->symbol) {
            if ($this->currencySymbol($currencyCode) == $symbol) {
                return $formater->formatCurrency($price, $currencyCode);
            } else {
                $formater->setSymbol(\NumberFormatter::CURRENCY_SYMBOL, $symbol);

                return $formater->format($this->convertPrice($price));
            }
        } else {
            return $formater->formatCurrency($price, $currencyCode);
        }
    }
` 
 paste above code in your facades.

Thanks

**Comment by farshadff** [comment]
@rahulshukla-webkul  thanks for your attention and your help , i edited to core function and paste your code but i still have the same problem prices disappear when i change the code in database when i saw your code i noticed that you are referring to a symbol which i dont have in my database table should i add it to currencies tables ???

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

Soon, we will give option to add currency symbol while creating currency and above code is only for removal 2 zeros. 

Thanks

**Comment by farshadff** [comment]
@rahulshukla-webkul dear rahul i replaced the function in core with the code you provided but still currencies have 2 seperators , should i do any thing else for that which i am not aware ?? thanks

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

Can you attach your **formatPrice** function code ?

**Comment by farshadff** [comment]
yes ofc so here is the path for insurance : `package/webkul/Core/src/Core.php` on line `422` is this core and function : 
```
  public function formatPrice($price, $currencyCode)
    {
        if (is_null($price))
            $price = 0;

        $formater = new \NumberFormatter( app()->getLocale(), \NumberFormatter::CURRENCY );

        $formater->setAttribute( $formater::FRACTION_DIGITS, 0 );

        if ($symbol = $this->getCurrentCurrency()->symbol) {
            if ($this->currencySymbol($currencyCode) == $symbol) {
                return $formater->formatCurrency($price, $currencyCode);
            } else {
                $formater->setSymbol(\NumberFormatter::CURRENCY_SYMBOL, $symbol);

                return $formater->format($this->convertPrice($price));
            }
        } else {
            return $formater->formatCurrency($price, $currencyCode);
        }
    }
```

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

Can you attach screenshot of price, which is coming on shop.

**Comment by farshadff** [comment]
@rahulshukla-webkul  yes ofc here is a picture of database and Frontside of shop of that product price
![product](https://user-images.githubusercontent.com/17511289/70893719-e3abd700-2000-11ea-94c6-02386f0cca79.jpg)
![asd](https://user-images.githubusercontent.com/17511289/70893726-e60e3100-2000-11ea-8a39-cb3045473d0b.jpg)

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

Can you give us dump of  your database and how you want it to display .

**Comment by farshadff** [comment]
yes @rahulshukla-webkul i want to show the price exacly as it enters for example if i enter like `100` it shows `100` not `100,00` because my currency doesnt have any seperators . and here is a dump of my database .

[dump.zip](https://github.com/bagisto/bagisto/files/3967141/dump.zip)
if you see some extra table its because i am adding some new features like submitting and activating the guarantee of product and so many many new features that i want to add to this product + many enhances in translate and some parts of code i am just trying to figure out a good way to commit them because it makes a bit nonsense as a feature or some thing

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

When we are using your database and with above function, we are getting - https://prnt.sc/qbhgzx

**Note -** We have only changed price from 200000 to 222222.

**Comment by farshadff** [comment]
@rahulshukla-webkul ok , thanks then i have to figure out the problem in my code . the weird thing is that even when i remove the function formatPrice nothing happens to the prices which is so weird to me . thanks for taking time and checking the things

**Comment by farshadff** [comment]
@rahulshukla-webkul i am sorry for activating this issue again because its so weird i installed a fresh bagisto like 4 times with this database 2 times it gave the save result as your image and 2 times the same as mine its so weird . when i `dd` anything in format price function it shows no where even when i delete this function nothing happens even on a fresh installation any idea what could be possibly wrong ?

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

Kindly push the same code on your fork branch, i will see from there and mention your branch name here so that we can resolve it.

**Comment by farshadff** [comment]
i have pushed to this repo with the name of branch `feature/final-show` but if you are not able to see that on my account there is a public repo named shopi have pushed my last code on that .

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

Kindly give github link.

**Comment by farshadff** [comment]
@rahulshukla-webkul  sorry i forgot the link :) 
```https://github.com/farshadff/shop```
 here you go thanks

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

We both have mismatched with version.
Actually i am answering according to current master branch and you are using 0.1.8.

In the same file , find a function named as - **public function currency($amount = 0)**
and replace the below code -

` /**
    * Format and convert price with currency symbol
    *
    * @param float $price
    *  @return string
    */
    public function currency($amount = 0)
    {
        if (is_null($amount))
            $amount = 0;

        $formatter = new \NumberFormatter( app()->getLocale(), \NumberFormatter::CURRENCY );

        $formatter->setAttribute( $formatter::FRACTION_DIGITS, 0 );

        return $formatter->formatCurrency($this->convertPrice($amount), $this->getCurrentCurrency()->code);
    }`

Basically - 

`$formatter->setAttribute( $formatter::FRACTION_DIGITS, 0 );`

is used to remove fraction digits from price and you can use it in different functions , if price is returning from others functions also.

Thanks

**Comment by farshadff** [comment]
@rahulshukla-webkul  thanks for being so much friendly and check it with patience , is there any way to pull the latest version if there are some good enhancement and some big differences in this 2 versions ?? thanks and soon i will put a good theme on this repo for others to use on bootstrap . :) thanks for your good effort

**Comment by farshadff** [comment]
inCase any one else would have this problem after this steps you would still have the seperator problem in admin dashboard which you must put this line : 
```
        $formatter->setAttribute( $formatter::FRACTION_DIGITS, 0 );
``` 
in the `formatBasePrice` function in core.php so you replace it with below code : 
```
   public function formatBasePrice($price)
    {
        if (is_null($price))
            $price = 0;

        $formatter = new \NumberFormatter( app()->getLocale(), \NumberFormatter::CURRENCY );
        $formatter->setAttribute( $formatter::FRACTION_DIGITS, 0 );

        return $formatter->formatCurrency($price, $this->getBaseCurrencyCode());
    }
```

**Comment by rahulshukla-webkul** [comment]
Hi @farshadff 

We are using separator that's why we did not use  -

`$formatter->setAttribute( $formatter::FRACTION_DIGITS, 0 );`.

You can pull master branch from our repository - https://github.com/bagisto/bagisto

Thanks
